_posts/2019-11-09-view-model-repository.md

```diff
@@ -0,0 +1,128 @@
+
+When you're used to work in MVC migrating to a different pattern might sound hard and you might give up before even trying. If you want to do MVC, but start using some cleaner code you can. It probably won't be as good as something written in VIPER, but not all apps need such big of an architecture.
+
+### MVC
+
+Even using MVC you can still use a ViewModel to separate your business logic and your presentation layer without having to use any reacting framework or data binding.
+
+```
+struct User {
+    let name: String
+    let surname: String
+}
+extension User: Codable {}
+```
+
+Let's assume we have a tableView and we want to display a list of users:
+
+```
+    private let cellReuseIdentifier = "reuseIdentifier"
+    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+         guard let cell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier, for: indexPath) as? UserTableViewCell else {
+            fatalError()
+        }
+        let model = model[indexPath.row]
+        cell.populate(withModel: model)
+        return cell
+    }
+```
+
+Every call to cellForRowAt indexPath should be this short. Configuring the cell should be done from the cell itself and not in the controller. Look for opportunities to move logic where it should be. When starting out with MVC we find ourselves to put everything in the controller without thinking where the right place for this function should be.
+
+We can now let the cell configure itself based on the model. Inside the cell class:
+
+```
+    func populate(withModel model: User) {
+        titleLabel.text = model.name
+        descriptionLabel.text = model.surname
+    }
+```
+
+### View Model
+
+We did not explain how we got the model object. In this case it is pretty likely that the model and the view model are the same since the object is really small and composed only of strings that map directly to fields in the view. If that's not the case we will need an intermediate layer that converts our model to a viewModel.
+
+We first define our model that maps to the object (some people call this entity) and the viewModel that directly maps to the various fields in the view. We also need to define how to convert from one to the other:
+
+```
+struct AmountModel {
+    let amount: Double
+}
+struct AmountViewModel {
+    let amount: String
+}
+extension AmountViewModel {
+    init(model: AmountModel, formatter: NumberFormatter) {//1
+        self.amount = formatter.string(from: model.amount as NSNumber) ?? ""
+    }
+}
+```
+
+1. Note how the view model doesn't know about the formatter, it gets told what formatter to use.
+
+Usually people using MVVM put all the logic inside the viewModel. I strongly disagree. The viewModel should be dumb, as the model, the should only know how to configure themselves. They can have helper methods, but they should not communicate directly to databases or initiate networking calls.
```

Comment:
the - > it
themselves - > itself

Reply:
Unless you by it to stay plural, but still the - > they

Reply:
Which means there's 3 they in there...

Reply:
mmm not so sure about this.
model and view model are 2 -> plural